### PR TITLE
StateProof: State proof support

### PR DESCRIPTION
**Pull request by winder**
Add StateProof support to go SDK.

### CODE REVIEW

File: types/stateproof.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+type FalconPublicKey [FalconPublicKeySize]byte
+
+// FalconVerifier implements the type Verifier interface for the falcon signature scheme.
+type FalconVerifier struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	PublicKey FalconPublicKey `codec:"k"`
+}
+
+// Signature represents a signature in the merkle signature scheme using falcon signatures as an underlying crypto scheme.
```
Comment (winder) [inline types/stateproof.go:142]:
```suggestion
// FalconSignatureStruct represents a signature in the merkle signature scheme using falcon signatures as an underlying crypto scheme.
```

**Comment by winder** [comment]
closes #361

**Review by algoidan — COMMENTED** [review]
Overall looks good. 
Have two small comments

**Comment by algoidan** [comment]
we should remove `client/v2/algod/getProof.go` file since it was renamed

File: client/v2/algod/algod.go [hunk 2]
```diff
@@ -80,8 +80,8 @@ func (c *Client) Block(round uint64) *Block {
 	return &Block{c: c, round: round}
 }
 
-func (c *Client) GetProof(round uint64, txid string) *GetProof {
-	return &GetProof{c: c, round: round, txid: txid}
+func (c *Client) GetTransactionProof(round uint64, txid string) *GetTransactionProof {
```
[the remark this answers is not in the document]
Comment (winder) [reply client/v2/algod/algod.go:83]: That's a query parameter, you can find it in `getTransactionProof.go`

File: types/stateproof.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	// StateProofBasic is our initial state proof setup. using falcon keys and subset-sum hash
+	StateProofBasic StateProofType = 0
+
+	// NumStateProofTypes is the max number of types of state proofs
+	// that we support.  This is used as an allocation bound for a map
+	// containing different stateproof types in msgpack encoding.
+	NumStateProofTypes int = 1
+)
+
+// GenericDigest is a digest that implements CustomSizeDigest, and can be used as hash output.
```
[the remark this answers is not in the document]
Comment (winder) [reply types/stateproof.go:22]: I'm not sure there's a benefit, it's not actually needed for any crypto operations that the SDK supports.

**Comment by algoidan** [comment]
just so we will not forget. We also need to add txn256 field. Do we have a different PR for it?

File: client/v2/common/models/transaction_state_proof.go~Stashed changes [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package models
```
Comment (Eric-Warehime) [inline client/v2/common/models/transaction_state_proof.go~Stashed changes:1]: This file seems unintended.

File: types/block.go [hunk 5]
```diff
@@ -93,6 +93,11 @@ type (
 		// started being supported).
 		TxnCounter uint64 `codec:"tc"`
```
Comment (algoidan) [inline types/block.go:94]: we should add the txn256 commitment here as well.
https://github.com/algorand/go-algorand/blob/7de989aa62a01ce9971b48cb1c2ca6cc2f18f3bf/data/bookkeeping/block.go#L141

File: types/stateproof.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	Proof                 SingleLeafProof `codec:"prf"`
+	VerifyingKey          FalconVerifier  `codec:"vkey"`
+}
+
+// A sigslotCommit is a single slot in the sigs array that forms the state proof.
+type sigslotCommit struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// Sig is a signature by the participant on the expected message.
+	Sig FalconSignatureStruct `codec:"s"`
```
Comment (algoidan) [inline types/stateproof.go:162]: This name is a little bit inaccurate. why not use the same name as in go-algorand? `MerkleSignature` or similar.

File: types/stateproof.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+// Message represents the message that the state proofs are attesting to. This message can be
+// used by lightweight client and gives it the ability to verify proofs on the Algorand's state.
+// In addition to that proof, this message also contains fields that
+// are needed in order to verify the next state proofs (VotersCommitment and LnProvenWeight).
+type Message struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+	// BlockHeadersCommitment contains a commitment on all light block headers within a state proof interval.
+	BlockHeadersCommitment []byte `codec:"b,allocbound=crypto.Sha256Size"`
```
Comment (algoidan) [inline types/stateproof.go:203]: There are lots of `allocbound` directives that use constants that are not defined.  Do we need to add it?

**Review by algoidan — COMMENTED** [review]
few more minor comments.

**Comment by Eric-Warehime** [comment]
Closing this in favor of #374

**Pull request closed without merging**